internal/ask/form/aggregation.go

```diff
@@ -0,0 +1,334 @@
+	MC    map[string]MCAggregation   `json:"mc" bson:"mc"`
+	Text  map[string]TextAggregation `json:"text" bson:"text"`
+}
+
+// Group defines a key for a multiple choice question / answer combo to be used
+// to define slices of submissions to be aggregated.
+type Group struct {
+	QuestionID string `json:"question_id" bson:"question_id"`
+	Question   string `json:"question" bson:"question"`
+	Answer     string `json:"answer" bson:"answer"`
+}
+
+//==============================================================================
+
+// AggregateFormSubmissions retrieves the submissions for a form, groups them then
+// runs aggregations and counts for each one.
+func AggregateFormSubmissions(context interface{}, db *db.DB, id string) (map[string]Aggregation, error) {
+
+	// Group the submissions.
+	groupedSubmissions, err := GroupSubmissions(context, db, id)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a container for the grouped aggregations.
+	groupAggregations := make(map[string]Aggregation)
+
+	// Loop through the groups of submissions.
+	for group, submissions := range groupedSubmissions {
+
+		// Perform the text aggregations.
+		textAggregations, err := TextAggregate(context, submissions)
+		if err != nil {
+			return nil, err
+		}
+
+		// Perform the multiple choice aggregations.
+		mcAggregations, err := MCAggregate(context, db, id, submissions)
+		if err != nil {
+			return nil, err
+		}
+
+		// Pack them in an aggregation along with the submission count and group.
+		agg := Aggregation{
+			Group: group,
+			Count: len(submissions),
+			MC:    mcAggregations,
+			Text:  textAggregations,
+		}
+
+		// Groups are ultimately based on a chosen answer. We do not have any keys for answers
+		// so hash the answer text for a unique key.
+		hasher := md5.New()
+		hasher.Write([]byte(group.Answer))
+		groupKey := hex.EncodeToString(hasher.Sum(nil))
+
+		// Add the aggregation to the map.
+		groupAggregations[groupKey] = agg
+
+	}
+
+	return groupAggregations, nil
+}
+
+//==============================================================================
+
+// GroupSubmissions organizes submissions by Group. It looks for questions with the group by flag
+// and creates Group structs.
+func GroupSubmissions(context interface{}, db *db.DB, id string) (map[Group][]submission.Submission, error) {
+
+	if !bson.IsObjectIdHex(id) {
+		log.Error(context, "TextAggregate", ErrInvalidID, "Completed")
+		return nil, ErrInvalidID
+	}
+
+	// Get the submissions for the form.Collection
+	subs, err := submission.Search(context, db, id, 0, 0, submission.SearchOpts{})
+	if err != nil {
+		return nil, err
+	}
+
+	groups := make(map[Group][]submission.Submission)
+
+	// Scan all the submissions and answers.
+	for _, sub := range subs.Submissions {
+
+		// Add all submissions to the [all,all] group
+		group := Group{
+			Question: "all",
+			Answer:   "all",
+		}
+		tmp := groups[group]
+		tmp = append(tmp, sub)
+		groups[group] = tmp
```

Comment:
This is a bit of an ugly requirement by Go. You cannot mutate a single element of a map due to provide safety in the way Go handles maps that may be accessed across threads.  To get around this I'm using a tmp var, mutating that, then setting it back to the map.